### PR TITLE
made checksum builder to resilient to missing data such as compilatio…

### PR DESCRIPTION
**Pull request by heejaechang**
…n/parse options which some projects such as typescript don't have.

also made creating solution to only include project type workspace knows about to create.

some of these will change once checksum moves into solution state itself, but the concept of having Null data will remain to support projects that don't have all data or filtering out projects that workspace doesn't know how to craete.

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @srivatsn @panopticoncentral @KevinH-MS @CyrusNajmabadi @dotnet/roslyn-ide can you take a look?

this is to unblock typescript.

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs [hunk 1]
```diff
@@ -89,8 +89,10 @@ private ChecksumCollection Build(IReadOnlyList<AnalyzerReference> key, IEnumerab
             var additionalDocuments = await subSnapshotBuilder.BuildAsync(projectState.AdditionalDocumentStates, projectState.AdditionalDocumentIds.Select(id => projectState.AdditionalDocumentStates[id]), WellKnownChecksumObjects.TextDocuments, cancellationToken).ConfigureAwait(false);
 
             var subAssetBuilder = new AssetBuilder(subTreeNode);
-            var compilationOptions = subAssetBuilder.Build(projectState, projectState.CompilationOptions, cancellationToken);
-            var parseOptions = subAssetBuilder.Build(projectState, projectState.ParseOptions, cancellationToken);
+
+            // set Asset.Nil if this particular project doesn't support compiler options
+            var compilationOptions = projectState.CompilationOptions != null ? subAssetBuilder.Build(projectState, projectState.CompilationOptions, cancellationToken) : Asset.Nil;
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: Should we just check `project.SupportsCompilation`?

File: src/Workspaces/Core/Portable/Execution/Asset.cs [hunk 2]
```diff
@@ -18,11 +18,29 @@ namespace Microsoft.CodeAnalysis.Execution
     /// </summary>
     internal abstract class Asset : ChecksumObject
     {
+        public static readonly Asset Nil = new NullAsset();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/Asset.cs]: "NullInstance"?

File: src/Workspaces/Core/Portable/Execution/Checksum.cs [hunk 3]
```diff
@@ -14,6 +14,8 @@ namespace Microsoft.CodeAnalysis.Execution
     /// </summary>
     internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
     {
+        public static readonly Checksum Nil = new Checksum(ImmutableArray<byte>.Empty);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/Checksum.cs]: Nil?  or Empty?

File: src/Workspaces/Core/Portable/Execution/ChecksumObject.cs [hunk 4]
```diff
@@ -73,6 +73,8 @@ private static Checksum CreateChecksum(string kind, object[] children)
     // TODO: Kind might not actually needed. see whether we can get rid of this
     internal static class WellKnownChecksumObjects
     {
+        public const string Nil = nameof(Nil);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumObject.cs]: why not Null?

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: sure, I can do that.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: agreed.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs]: hmm... there is no support syntax tree in project state. but yet parse option is there. strange.

File: src/Workspaces/Core/Portable/Execution/ChecksumObject.cs (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumObject.cs]: nil looks like include both null or empty or invalid. but sure can change everything to null.

File: src/Workspaces/Core/Portable/Execution/Asset.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Asset.cs]: just Null seems better.

**Comment by heejaechang** [comment]
okay to checkin?

File: src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs [hunk 5]
```diff
@@ -493,6 +495,23 @@ public class SnapshotSerializationTests : SnapshotSerializationTestBase
...
         }
 
+        [Fact]
+        public async Task UnknownLanguageTest()
+        {
+            var hostServices = MefHostServices.Create(MefHostServices.DefaultAssemblies.Add(typeof(NullLanguageService).Assembly));
+
+            var solution = new AdhocWorkspace(hostServices).CurrentSolution;
+
+            var project1 = solution.AddProject("Project", "Project.dll", NullLanguageService.TestLanguage);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs]: I believe we already have a language for this.  We have tests for this. look for NoCompilation

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs]: i.e.

```
                    <Project Language="NoCompilation">
                        <Document FilePath="A.ts">
                            Dummy content.
                        </Document>
                    </Project>
```

Comment (heejaechang) [inline src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs]: that exist for editor layer not service layer

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs]: We could always move that down then :)

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/Execution/SnapshotSerializationTests.cs]: I'd prefer to avoid duplication in test infrastructure.

**Comment by CyrusNajmabadi** [comment]
👍 if you're very busy.  But if you can do the test change, that would be appreciated so we can have test consistency.  I view test code as just as important as product code.  So we should apply the same vigilance there.

**Comment by heejaechang** [comment]
I will do that in my next PR if I can - https://github.com/dotnet/roslyn/pull/13773 - which i am moving bunch of stuff from Editor unit test to Service Utilities to share our test MEF exports.

moving MEF (especially test MEF type) is not simple. require changing a lot of code around.

**Pull request merged**